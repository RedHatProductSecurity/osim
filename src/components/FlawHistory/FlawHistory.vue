<script setup lang="ts">
import { computed, ref } from 'vue';

import { usePagination } from '@/composables/usePagination';

import EditableDate from '@/widgets/EditableDate/EditableDate.vue';
import { capitalize, formatDateWithTimezone } from '@/utils/helpers';
import { flawFieldNamesMapping } from '@/constants/flawFields';
import type { ZodFlawHistoryItemType } from '@/types/zodFlaw';
import LabelCollapsible from '@/widgets/LabelCollapsible/LabelCollapsible.vue';

const props = defineProps<{
  history: null | undefined | ZodFlawHistoryItemType[];
}>();

const startDate = ref<null | string | undefined>(null);
const endDate = ref<null | string | undefined>(null);

const emptyFilters = computed(() => {
  return !startDate.value && !endDate.value;
});

const validDateRange = computed(() => {
  const start = startDate.value ? new Date(startDate.value) : null;
  const end = endDate.value ? new Date(endDate.value) : null;

  return (
    start instanceof Date && !Number.isNaN(start.getTime())
    && end instanceof Date && !Number.isNaN(end.getTime())
    && end >= start
  );
});

const filteredHistoryItems = computed(() => {
  if (!validDateRange.value) {
    return props.history;
  }

  const start = new Date(startDate.value!);
  const end = new Date(endDate.value!);
  end.setDate(end.getDate() + 1);

  return props.history?.filter((item) => {
    if (!item.pgh_created_at) return false;

    const itemDate = new Date(item.pgh_created_at);

    return itemDate.getTime() >= start.getTime() && itemDate.getTime() <= end.getTime();
  });
});

const historyExpanded = ref(true);

const itemsPerPage = 20;
const totalPages = computed(() =>
  Math.ceil((filteredHistoryItems.value?.length || 0) / itemsPerPage),
);

const {
  changePage,
  currentPage,
  pages,
} = usePagination(totalPages, itemsPerPage);

const paginatedHistoryItems = computed(() => {
  const start = (currentPage.value - 1) * itemsPerPage;
  const end = start + itemsPerPage;
  return filteredHistoryItems.value?.slice(start, end);
});

function isDateField(field: string) {
  return field.includes('_dt');
}

function clearFilters() {
  startDate.value = null;
  endDate.value = null;
}
</script>

<template>
  <LabelCollapsible
    class="my-2"
    :class="{'pb-4': !history?.length}"
    :isExpanded="historyExpanded"
    @toggleExpanded="historyExpanded = !historyExpanded"
  >
    <template #label>
      <label class="mx-2 mb-0 form-label">
        <h4 :class="{'mb-0': !historyExpanded}">History</h4>
      </label>
    </template>
    <template v-if="!history?.length">
      <span>There are no tracked changes for this flaw.</span>
    </template>
    <template v-else>
      <div class="d-flex mb-2 gap-2">
        <button
          tabindex="-1"
          type="button"
          :disabled="emptyFilters"
          class="input-group-text"
          @click="clearFilters()"
          @mousedown="event => event.preventDefault()"
        >
          <i class="bi bi-eraser"></i>
        </button>
        <EditableDate
          v-model="startDate as string"
          style="width: 225px;"
          placeholder="[Start date]"
        />
        <EditableDate
          v-model="endDate as string"
          style="width: 225px;"
          placeholder="[End date]"
        />
      </div>
      <template v-if="!filteredHistoryItems?.length">
        <span>There are no results for current filter.</span>
      </template>
      <div v-else class="mt-2">
        <div class="mt-2">
          <template v-for="historyEntry in paginatedHistoryItems" :key="historyEntry.pgh_slug">
            <div v-if="historyEntry.pgh_diff" class="alert alert-info mb-1 p-2">
              <span>
                {{ formatDateWithTimezone(historyEntry.pgh_created_at || '', true) }}
                - {{ historyEntry.pgh_context?.user || 'System' }}
              </span>
              <ul class="mb-2">
                <li v-for="(diffEntry, diffKey) in historyEntry.pgh_diff" :key="diffKey">
                  <div class="ms-3 pb-0">
                    <span>{{ capitalize(historyEntry.pgh_label) }}</span>
                    <span class="fw-bold">{{ ' ' + (flawFieldNamesMapping[diffKey] || diffKey) }}</span>
                    <span>{{ ': ' +
                      (isDateField(diffKey) && diffEntry[0]
                        ? formatDateWithTimezone(diffEntry[0])
                        : (diffEntry[0]?.toString() || '')
                      ) + ' '
                    }}</span>
                    <i class="bi bi-arrow-right" />
                    {{ (isDateField(diffKey) && diffEntry[1]
                      ? formatDateWithTimezone(diffEntry[1])
                      : (diffEntry[1]?.toString() || '')
                    ) }}
                  </div>
                </li>
              </ul>
            </div>
          </template>
        </div>
      </div>
    </template>
  </LabelCollapsible>
  <div v-if="history?.length" class="pagination-controls gap-1 my-2">
    <button
      type="button"
      tabindex="-1"
      class="btn btn-sm btn-secondary rounded-end-0"
      :disabled="currentPage === 1"
      @click="changePage(currentPage - 1)"
    >
      <i class="bi bi-arrow-left fs-5" />
    </button>
    <button
      v-for="page in pages"
      :key="page"
      tabindex="-1"
      class="osim-page-btn btn btn-sm rounded-0 btn-secondary"
      style="width: 34.8px;"
      :disabled="page === currentPage || page === '..'"
      @click.prevent="changePage(page as number)"
    >
      {{ page }}
    </button>
    <button
      type="button"
      tabindex="-1"
      class="btn btn-sm btn-secondary rounded-start-0"
      :disabled="currentPage === totalPages || totalPages === 0"
      @click.prevent="changePage(currentPage + 1)"
    >
      <i class="bi bi-arrow-right fs-5" />
    </button>
  </div>
</template>

<style scoped lang="scss">
    .pagination-controls {
  display: flex;

  button {
    height: 2rem;
    padding-block: 0;

    &.osim-page-btn:disabled {
      background-color: transparent;
      color: black;
    }
  }
}
</style>
