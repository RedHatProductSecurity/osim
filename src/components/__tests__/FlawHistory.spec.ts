import { computed, ref, type Directive } from 'vue';

import { mount } from '@vue/test-utils';
import { IMaskDirective } from 'vue-imask';

import FlawHistory from '@/components/FlawHistory/FlawHistory.vue';

import type { ZodFlawHistoryItemType } from '@/types/zodFlaw';

import { osimEmptyFlawTest, osimFullFlawTest } from './test-suite-helpers';

function sampleHistoryItem(): ZodFlawHistoryItemType {
  return {
    pgh_created_at: '2024-10-04T15:06:56.760289Z',
    pgh_slug: 'osidb.FlawAudit:123456',
    pgh_label: 'update',
    pgh_context: { url: 'osidb/api/v1/flaws/12d3820a-a43b-417c-a22e-46e47c232a63', user: 1 },
    pgh_diff: { owner: ['', 'noemail@example.com'] },
  };
}

describe('flawHistory', () => {
  osimEmptyFlawTest('is not shown if no history present on flaw', async ({ flaw }) => {
    const subject = mount(FlawHistory, {
      props: {
        history: flaw.history,
      },
      global: {
        stubs: {
          EditableDate: true,
        },
      },
    });
    const historyItems = subject.findAll('.alert-info');
    expect(historyItems.length).toBe(0);
  });

  osimFullFlawTest('is shown if history present on flaw', async ({ flaw }) => {
    flaw.history = [];
    flaw.history.push(sampleHistoryItem());
    const subject = mount(FlawHistory, {
      props: {
        history: flaw.history,
      },
      global: {
        stubs: {
          EditableDate: true,
        },
      },
    });
    const historyItems = subject.findAll('.alert-info');
    expect(historyItems.length).toBe(1);
    expect(subject.html()).toMatchSnapshot();
  });

  osimFullFlawTest('displays history items', async ({ flaw }) => {
    flaw.history = [];
    flaw.history.push(sampleHistoryItem());
    const subject = mount(FlawHistory, {
      props: {
        history: flaw.history,
      },
      global: {
        stubs: {
          EditableDate: true,
        },
      },
    });
    const historyListItem = subject.find('li div');
    expect(historyListItem?.text()).includes('Update Owner:');
  });

  osimFullFlawTest('shows seconds, minutes, hours in timestamp', async ({ flaw }) => {
    flaw.history = [];
    flaw.history.push(sampleHistoryItem());
    const subject = mount(FlawHistory, {
      props: {
        history: flaw.history,
      },
      global: {
        stubs: {
          EditableDate: true,
        },
      },
    });
    const historyEntry = subject.find('div.alert-info');
    expect(historyEntry?.text()).includes('15:06:56');
  });

  it('filters history items based on date range', async () => {
    const mockHistory = [{
      pgh_created_at: '2024-10-04T15:00:00.000000Z',
      pgh_slug: 'osidb.FlawAudit:123456',
      pgh_label: 'update',
      pgh_context: { url: 'osidb/api/v1/flaws/12d3820a-a43b-417c-a22e-46e47c232a63', user: 1 },
      pgh_diff: { owner: ['', 'noemail@example.com'] },
    }, {
      pgh_created_at: '2024-10-14T15:08:46.760289Z',
      pgh_slug: 'osidb.FlawAudit:123456',
      pgh_label: 'update',
      pgh_context: { url: 'osidb/api/v1/flaws/12d3820a-a43b-417c-a22e-46e47c232a63', user: 1 },
      pgh_diff: { owner: ['', 'noemail@example.com'] },
    }];

    const props = {
      history: mockHistory,
    };

    const startDate = ref('');
    const endDate = ref('');
    const validDateRange = computed(() => !!startDate.value && !!endDate.value);

    const filteredHistoryItems = computed(() => {
      if (!validDateRange.value) {
        return props.history;
      }

      const start = new Date(startDate.value!);
      const end = new Date(endDate.value!);
      end.setDate(end.getDate() + 1);

      return props.history?.filter((item) => {
        if (!item.pgh_created_at) return false;

        const itemDate = new Date(item.pgh_created_at);

        return itemDate.getTime() >= start.getTime() && itemDate.getTime() <= end.getTime();
      });
    });

    const subject = mount(FlawHistory, {
      props,
      global: {
        provide: {
          startDate,
          endDate,
          validDateRange,
        },
        directives: {
          imask: IMaskDirective as Directive,
        },
      },
    });

    expect(filteredHistoryItems.value).toEqual(mockHistory);

    startDate.value = '2024-10-03';
    endDate.value = '2024-10-04';
    await subject.vm.$nextTick();
    expect(filteredHistoryItems.value.length).toEqual(1);

    startDate.value = '2024-02-01';
    endDate.value = '2024-02-28';
    await subject.vm.$nextTick();
    expect(filteredHistoryItems.value).toEqual([]);
  });
});
