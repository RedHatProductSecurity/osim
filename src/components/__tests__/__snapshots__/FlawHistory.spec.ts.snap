// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`flawHistory > is shown if history present on flaw 1`] = `
"<div data-v-4715e8e2="" data-v-f4fcc8d4="" class="osim-collapsible-label my-2"><button data-v-4715e8e2="" type="button" class="me-2 osim-collapsible-toggle"><i data-v-4715e8e2="" class="bi me-1 bi-dash-square-dotted"></i><label data-v-f4fcc8d4="" class="mx-2 mb-0 form-label">
      <h4 data-v-f4fcc8d4="" class="">History</h4>
    </label></button>
  <div data-v-4715e8e2="" class="ps-3 border-start">
    <div data-v-4715e8e2="" class="">
      <div data-v-f4fcc8d4="" class="d-flex mb-2 gap-2"><button data-v-f4fcc8d4="" tabindex="-1" type="button" disabled="" class="input-group-text"><i data-v-f4fcc8d4="" class="bi bi-eraser"></i></button>
        <editable-date-stub data-v-f4fcc8d4="" placeholder="[Start date]" editing="false" includestime="false" readonly="false" style="width: 225px;"></editable-date-stub>
        <editable-date-stub data-v-f4fcc8d4="" placeholder="[End date]" editing="false" includestime="false" readonly="false" style="width: 225px;"></editable-date-stub>
      </div>
      <div data-v-f4fcc8d4="" class="mt-2">
        <div data-v-f4fcc8d4="" class="mt-2">
          <div data-v-f4fcc8d4="" class="alert alert-info mb-1 p-2"><span data-v-f4fcc8d4="">2024-10-04 15:06:56 UTC - 1</span>
            <ul data-v-f4fcc8d4="" class="mb-2">
              <li data-v-f4fcc8d4="">
                <div data-v-f4fcc8d4="" class="ms-3 pb-0">
                  <!--v-if--><span data-v-f4fcc8d4="">Update</span><span data-v-f4fcc8d4="" class="fw-bold"> Owner</span><span data-v-f4fcc8d4="">:  </span><i data-v-f4fcc8d4="" class="bi bi-arrow-right"></i> noemail@example.com
                </div>
              </li>
            </ul>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
<div data-v-f4fcc8d4="" class="pagination-controls gap-1 my-2"><button data-v-f4fcc8d4="" type="button" tabindex="-1" class="btn btn-sm btn-secondary rounded-end-0" disabled=""><i data-v-f4fcc8d4="" class="bi bi-arrow-left fs-5"></i></button><button data-v-f4fcc8d4="" tabindex="-1" class="osim-page-btn btn btn-sm rounded-0 btn-secondary" style="width: 34.8px;" disabled="">1</button><button data-v-f4fcc8d4="" type="button" tabindex="-1" class="btn btn-sm btn-secondary rounded-start-0" disabled=""><i data-v-f4fcc8d4="" class="bi bi-arrow-right fs-5"></i></button></div>"
`;
